### PR TITLE
Ag 5160/charts legend

### PR DESCRIPTION
**Pull request by manapeirov**
This PR is for https://ag-grid.atlassian.net/browse/AG-5160 and https://ag-grid.atlassian.net/browse/AG-6852.

The aim is to allow the user to specify a max width for the legend items in order to prevent large items from increasing the overall dynamic width of the legend. The previous logic which hides the legend if it exceeds 50% of the available size is still present. Specifying a maxWidth on the legend item options will prevent the legend from squashing the chart and being hidden. 

An option has also been added to the control the character limit of the legend label text.

### CODE REVIEW

File: charts-packages/ag-charts-community/src/chart/legend.ts [hunk 1]
```diff
@@ -202,24 +211,64 @@ export class Legend {
...
+                addEllipsis = true;
+            }
+
+            if (maxWidth) {
+                const labelBBox = markerLabel.computeBBox();
+                if (labelBBox.width > maxWidth!) {
+                    let truncatedText = '';
+                    let cumCharSize = characterSizeMap[ellipsis];
+
+                    for (let char of textChars) {
```
Comment (alantreadway) [inline charts-packages/ag-charts-community/src/chart/legend.ts:246]: Nit: Prefer `const` here if possible?

File: charts-packages/ag-charts-community/src/chart/legend.ts [hunk 2]
```diff
@@ -202,24 +211,64 @@ export class Legend {
...
-            markerLabel.fontWeight = itemLabel.fontWeight;
-            markerLabel.fontSize = itemLabel.fontSize;
-            markerLabel.fontFamily = itemLabel.fontFamily;
-
             let text = datum.label.text;
-            if (text.length > maxCharCount) {
-                text = `${text.substring(0, maxCharCount - ellipsis.length)}${ellipsis}`;
-            }
             markerLabel.text = text;
+            markerLabel.markerSize = marker.size;
```
[the remark this answers is not in the document]
Comment (manapeirov) [reply charts-packages/ag-charts-community/src/chart/legend.ts]: done

File: charts-packages/ag-charts-community/src/chart/agChartOptions.ts [hunk 3]
```diff
@@ -288,6 +288,8 @@ export interface AgChartLegendMarkerOptions {
 }
 
 export interface AgChartLegendLabelOptions {
+    /** If the label text exceeds the character limit, it will be truncated and an ellipsis will be appended to indicate this. By default, the character limit is 25. */
```
[the remark this answers is not in the document]
Comment (manapeirov) [reply charts-packages/ag-charts-community/src/chart/agChartOptions.ts:291]: done

File: charts-packages/ag-charts-community/src/chart/legend.ts [hunk 4]
```diff
@@ -202,24 +211,64 @@ export class Legend {
...
+                    }
+
+                    text = truncatedText;
+                    addEllipsis = true;
+                }
+            }
+
+            if (addEllipsis) {
+                text += ellipsis;
+                markerLabel.text = text;
```
[the remark this answers is not in the document]
Comment (manapeirov) [reply charts-packages/ag-charts-community/src/chart/legend.ts]: fixed

File: charts-packages/ag-charts-community/src/chart/legend.ts [hunk 5]
```diff
@@ -202,24 +211,64 @@ export class Legend {
...
+            let addEllipsis = false;
+
+            if (text.length > characterLimit) {
+                text = `${text.substring(0, characterLimit - ellipsis.length)}`;
+                addEllipsis = true;
+            }
+
+            if (maxWidth) {
+                const labelBBox = markerLabel.computeBBox();
+                if (labelBBox.width > maxWidth!) {
```
[the remark this answers is not in the document]
Comment (manapeirov) [reply charts-packages/ag-charts-community/src/chart/legend.ts]: It was giving an error before during build but seems to be ok now

File: charts-packages/ag-charts-community/src/chart/chart.ts [hunk 6]
```diff
@@ -879,6 +879,8 @@ export abstract class Chart extends Observable {
         const height = this.height - captionAutoPadding;
         const legendGroup = legend.group;
         const legendSpacing = legend.spacing;
+        const legendItemMaxWidthPercentage = 0.8;
+        legend.maxItemWidth = legend.item.maxWidth ?? (width * legendItemMaxWidthPercentage);
```
Comment (alantreadway) [inline charts-packages/ag-charts-community/src/chart/chart.ts:883]: Could we move the responsibility for this into the `Legend` class? It doesn't seem to relate to the `Chart` at all?

**Review by alantreadway — COMMENTED** [review]
Looking much better - thanks! One outstanding query regarding location of the defaulting logic.

File: charts-packages/ag-charts-community/src/chart/legend.ts [hunk 7]
```diff
@@ -180,6 +180,21 @@ export class Legend {
      */
     spacing = 20;
 
+    private characterWidths = new Map();
+
+    private getCharacterWidths(font: string) {
+        const { characterWidths } = this;
+        if (characterWidths.has(font)) {
+            return characterWidths.get(font);
+        } else {
```
Comment (alantreadway) [inline charts-packages/ag-charts-community/src/chart/legend.ts:189]: Nit: Drop unnecessary `else` block (outdent block contents by 1).

File: charts-packages/ag-charts-community/src/chart/chart.ts (hunk 6 above)
Comment (manapeirov) [reply charts-packages/ag-charts-community/src/chart/chart.ts]: done

**Review by alantreadway — APPROVED** [review]
LGTM!

**Pull request merged**